### ISSUE TITLE
用fp32精度微调生成的模型过大

### ISSUE THREAD

**Issue by zhenhao-huang** [TsinghuaAI/CPM-Finetune#16]
在启用了`fp32`精度训练之后，即使用了`finetune_chid_large_fp32.sh`脚本里面的参数微调chid，每次保存的模型至少有28G之大，在我的数据集上生成的模型也是。

**Comment by zhenhao-huang** [comment]
微调过程loss为0是什么情况？

**Comment by t1101675** [comment]
1. 模型较大的问题是因为 fp32 保存的模型里面还包含了 optimizer 的 state 等其他信息
2. 能否提供一下 loss 为 0 的运行时 log？

**Comment by zhenhao-huang** [comment]
我是在自己的模板上进行文本生成的微调，1个epoch的loss日志如下：

> Train losses:
Epoch 0, global step 10, total step 41003, train lm loss: 21.50133409500122
Epoch 0, global step 20, total step 41003, train lm loss: 21.995201778411865
Epoch 0, global step 30, total step 41003, train lm loss: 22.658365488052368
Epoch 0, global step 40, total step 41003, train lm loss: 22.105499076843262
Epoch 0, global step 50, total step 41003, train lm loss: 20.980561780929566
Epoch 0, global step 60, total step 41003, train lm loss: 19.740772199630737
Epoch 0, global step 70, total step 41003, train lm loss: 18.69202947616577
Epoch 0, global step 80, total step 41003, train lm loss: 16.577645540237427
Epoch 0, global step 90, total step 41003, train lm loss: 14.256615781784058
Epoch 0, global step 100, total step 41003, train lm loss: 10.281423848867416
Epoch 0, global step 110, total step 41003, train lm loss: 2.9970952646806834
Epoch 0, global step 120, total step 41003, train lm loss: 0.22705000592391117
Epoch 0, global step 130, total step 41003, train lm loss: 9.655901942551282e-05
Epoch 0, global step 140, total step 41003, train lm loss: 1.738655853387172e-05
Epoch 0, global step 150, total step 41003, train lm loss: 1.4781899608351523e-06
Epoch 0, global step 160, total step 41003, train lm loss: 5.006785158911953e-07
Epoch 0, global step 170, total step 41003, train lm loss: 9.240420308742613e-06
Epoch 0, global step 180, total step 41003, train lm loss: 5.685963401447225e-06
Epoch 0, global step 190, total step 41003, train lm loss: 0.10254945754752498
Epoch 0, global step 200, total step 41003, train lm loss: 0.0002156142704279773
...
Epoch 0, global step 960, total step 41003, train lm loss: 2.4735838906053685e-07
Epoch 0, global step 970, total step 41003, train lm loss: 0.0
Epoch 0, global step 980, total step 41003, train lm loss: 1.2814969068131178e-07
Epoch 0, global step 990, total step 41003, train lm loss: 2.890810264943866e-07
Epoch 0, global step 1000, total step 41003, train lm loss: 1.0073015801026486e-06
Epoch 0, global step 1010, total step 41003, train lm loss: 6.437223873945186e-07
Epoch 0, global step 1020, total step 41003, train lm loss: 2.0563522866723362e-07
Epoch 0, global step 1030, total step 41003, train lm loss: 8.046615675993963e-08
Epoch 0, global step 1040, total step 41003, train lm loss: 3.5762657262239373e-07
Epoch 0, global step 1050, total step 41003, train lm loss: 1.7583329281478655e-07
Epoch 0, global step 1060, total step 41003, train lm loss: 4.470348358154297e-08
Epoch 0, global step 1070, total step 41003, train lm loss: 2.384185791015625e-08
Epoch 0, global step 1080, total step 41003, train lm loss: 5.323327427504409e-05
Epoch 0, global step 1090, total step 41003, train lm loss: 2.9802322387695314e-09
Epoch 0, global step 1100, total step 41003, train lm loss: 0.0
Epoch 0, global step 1110, total step 41003, train lm loss: 3.6656601878348736e-07
Epoch 0, global step 1120, total step 41003, train lm loss: 0.0
Epoch 0, global step 1130, total step 41003, train lm loss: 2.9802322387695314e-09
...
Epoch 0, global step 4370, total step 41003, train lm loss: 2.9802322387695314e-09
Epoch 0, global step 4380, total step 41003, train lm loss: 0.0
Epoch 0, global step 4390, total step 41003, train lm loss: 0.0
Epoch 0, global step 4400, total step 41003, train lm loss: 0.0
Epoch 0, global step 4410, total step 41003, train lm loss: 0.0
Epoch 0, global step 4420, total step 41003, train lm loss: 0.0
Epoch 0, global step 4430, total step 41003, train lm loss: 1.4901161193847656e-08
Epoch 0, global step 4440, total step 41003, train lm loss: 0.0
Epoch 0, global step 4450, total step 41003, train lm loss: 0.0
Epoch 0, global step 4460, total step 41003, train lm loss: 0.0
Epoch 0, global step 4470, total step 41003, train lm loss: 0.0
Epoch 0, global step 4480, total step 41003, train lm loss: 1.4901161193847656e-08
Epoch 0, global step 4490, total step 41003, train lm loss: 0.0
Epoch 0, global step 4500, total step 41003, train lm loss: 0.0
Epoch 0, global step 4510, total step 41003, train lm loss: 5.960464477539063e-09
Epoch 0, global step 4520, total step 41003, train lm loss: 2.9802322387695314e-09
...

是我的`loss_mask`计算方式不对吗，我参考了`finetune_chid`计算`loss_mask`的方法：`loss_mask = [0] * (len(d["sent"]) - 2) + [1]`，`loss`计算也是参考了`finetune_chid`的方法

**Comment by t1101675** [comment]
不太清楚您具体的任务设定是什么样的。ChID 的模板里面，一句话只有最后一个词上面会有 loss，对应着代码里面 loss_mask 的构造方法（最后一位为1，前面都为0）。如果做文本生成的话，我理解应该所有词上面都需要有 loss，所以 loss_mask 应该都是 1 才对。不知道是不是这个问题

**Comment by zhenhao-huang** [comment]
我做的是短文本生成，类似于对话生成。修改成`loss_mask = [1] * (len(d) - 1)`之后，目前`loss`没出现`’0‘`这种异常，感谢

**Comment by LaVineChan** [comment]
> 我做的是短文本生成，类似于对话生成。修改成`loss_mask = [1] * (len(d) - 1)`之后，目前`loss`没出现`’0‘`这种异常，感谢

请问一下您，您的这个`loss_mask = [1] * (len(d) - 1)`是不是相当于你的数据为上文: paragraph1 下文: paragraph2，计算loss 的时候计算的是上文: paragraph1 下文: paragraph2这一完整输入数据，而不是只考虑生成的结果 paragraph2。因为我现在做问题生成，数据形式为(context : answer : question), 我想我这种场景是需要只计算question的对应位置的loss，还是所有完整输入数据的loss？

**Comment by zhenhao-huang** [comment]
@LaVineChan 是的，我的`loss`计算的是完整输入，您的场景应该也是需要计算完整输入数据的`loss`，可以问问作者的建议
@t1101675

**Comment by t1101675** [comment]
我们其实建议只计算 question loss 的输入。因为在我们之前的一些实验里面，我们发现将一些无关的位置 mask 掉可以提升模型训练的收敛速度。
